File: convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun sortedArrayToBST(nums: IntArray, s:Int = 0, e: Int = nums.size): TreeNode? {
        if(s >= e) return null
        var m = (s+e)/2
        var ret = TreeNode(nums[m])
        ret.left = sortedArrayToBST(nums, s, m)
        ret.right = sortedArrayToBST(nums,m+1,e)
        return ret
    }
}

File: convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums, int i, int j) {
        if(i == j) return 0;
        int m = (i+j)/2;
        return new TreeNode(nums[m], sortedArrayToBST(nums,i,m), sortedArrayToBST(nums,m+1,j));
    }
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return sortedArrayToBST(nums, 0, nums.size());
    }
};